### ISSUE TITLE
How do you get a hit result from the Add and Fire bullet blueprint function?

### ISSUE THREAD

**Issue by vallamost** [ErikHedberg/Terminal-Ballistics-Example-Project#3]
As seen in the image, there's no hit result output on this function. I didn't see any documentation for the 'id' output it has. What are these used for if you can't get a hit result from them?

Are the callback function nodes the only nodes that contain hit information? 

![image](https://github.com/ErikHedberg/Terminal-Ballistics-Example-Project/assets/25071166/8f824bef-d6fc-4968-9851-13bbb480bb7b)

**Comment by wupy36** [comment]
you can use the id information, but I would suggest using the with callbacks, just don't use the callbacks with updates as it can be very expensive. 
![image](https://github.com/ErikHedberg/Terminal-Ballistics-Example-Project/assets/5973921/fd956639-ede9-4bf4-b9f1-fb912efaf3a6)

**Comment by ErikHedberg** [comment]
Additionally, if your game mode class implements the interface `ITerminalBallisticsGameModeBaseInterface`,
you will be able to implement events which are automatically called when a bullet hits something, injures someone, or exits something.
![image](https://github.com/ErikHedberg/Terminal-Ballistics-Example-Project/assets/13501941/fd84dbaf-24fc-4dbd-a486-908c0d8109c3)

**Comment by vallamost** [comment]
Thank you for that! For reference for others in the future, I am trying to integrate this with a water plugin in UE5 where bullets will make impacts on the water. In other to do that I need to fire projectiles that have colliders / collision with simulated physics on them.

In this case, am I limited to only the the "Add and Fire Projectile with Callbacks"? That is described in the wiki as having a physics collider on it, the bullets seem to be more like line trace objects. Is there an optimized way to make the Fire Bullet callback also shoot a physics actor with it? I don't need the bullets to be long lived. Is it as easy as assigning a static mesh to the DA_Bullet_* data assets and making sure that mesh has collision on it?

I've tried putting meshes in those data assets with collision on them but I don't see the meshes getting spawned when the bullets are fired nor do I see water impacts.

This is my test (ignore the incorrect 556 bullet casing size 😅 )
![image](https://github.com/ErikHedberg/Terminal-Ballistics-Example-Project/assets/25071166/f2990591-1cea-4ce9-bef2-594556135212)

I added the TerminalBallisticsGameModeBaseInterface to the Game Mode class as well to the Character BP class the player is using.

As seen in this screenshot below, my first debug print statement is fired, but the print statement for the On Bullet Hit event is not being triggered. These are in the Character blueprint event graph. Do I have to call the On Bullet hit event from the GameMode blueprint or is it fine to use them from the character blueprints? In the TB Example project they look to be coming from the character blueprint like what I have shown below but my Hit event isn't getting triggered. Maybe a collision channel is wrong?

Sorry for all the questions :(

Any help would be greatly appreciated!

Hit event wired up but not triggering. My  first debug statement is printed but the other one of "not triggering" from the hit event doesn't show up on the fire event.
![image](https://github.com/ErikHedberg/Terminal-Ballistics-Example-Project/assets/25071166/79a24fbb-2213-4e07-9a20-13f23fc81e31)

**Comment by vallamost** [comment]
I think I've narrowed down the issue but I haven't solved it yet, the Default Physics material isn't getting applied to objects for some reason as I found that the Bullet Complete event triggers a print message successfully I double checked and made sure I have a Projectile Object channel created with the Default Response set to Block in the Project Settings > Engine > Collision. I also have Physical Surface types created in the Engine > Physics settings.

I noticed in the Terminal Ballistic's Plugin Settings the Projectile Trace Channel was wrong as it inherited a channel already created. I corrected that but still no luck. I even migrated and imported the Terminal Ballistic test map of objects that have physical materials applied to them but hits don't register. 

I'm not sure what else to check at this time..

**Comment by vallamost** [comment]
Ahhh, I think I solved it, the make Launch Params was using the wrong trace channel, the project I was working on had a trace channel also called "Projectile", that was initially selected in the TB Launch Params until I realized it was the wrong one as it was only set to Overlap and not Block. Once I renamed that and remade the correct "Projectile" channel and selected it in the Launch Params the OnHit immediately started working. 

If there's any suggestions or answers on how to make the physics collider start working from the Fire Bullet action for water, that would  be appreciated!

Maybe I need to build out some logic that creates a physics impulse at the bullet hit location using information from the On Hit event?

One other issue I'm seeing is a print statement from TB that says "Bullet impact event triggered from GM", I'm guessing GM is gun muzzle? I have the bullets being spawned well in front of the muzzle, I am also using a line trace so I can see where they start and they aren't close to the gun so I don't know how that's possible. Ignore Owner is also checked in the Launch Params :|

**Comment by wupy36** [comment]
Have you setup the collision on the static mesh? 
- Collision Preset: BlockAll 
- Collision Complexity: Project Default
 
![image](https://github.com/ErikHedberg/Terminal-Ballistics-Example-Project/assets/5973921/1375b8bf-6ca5-4111-a9d2-222d2033389f)


What water plugin are you using?

**Comment by vallamost** [comment]
> Have you setup the collision on the static mesh?
> 
>     * Collision Preset: BlockAll
> 
>     * Collision Complexity: Project Default
> 
> 
> 

Sorry if it wasn't clear, the collision on static meshes like walls are working now but the projectile / bullet collision isn't affecting the water as I suspect it isn't launching a short lived physics collider at them. The weapon I am using discharges a bullet casing that has a static mesh and collider, that affects the water just fine. 

To make troubleshooting easier, I switched my gun skeletal mesh to the one from the TB example project (SK_FPGun) with the same Bone name for the fire location and fire direction but I am still seeing really weird issues as seen in this recording I took:

https://youtu.be/omFeA3cCxi4

I used the same Get Socket Transform location and rotation for the Launch Params for the SK_FPGun asset but I am getting funky results, like bullets curving in midair. It's also registering objects being hit that are nowhere near the debug line traces, quite odd..

> What water plugin are you using?

I am using Waterline Pro.

**Comment by ErikHedberg** [comment]
You are correct, no physics bodies are being created and launched. That is intentional.
With regards to the bullets appearing to "curve", based on your video it looks like you are shooting inside of or close to a water volume, and it seems that your "water volume activation box" is blocking the projectiles. It is likely that the bullets are bouncing off the water or experiencing extreme changes in velocity due to them already being inside the water. In addition, some of the shots that hit the pillar were clearly not close enough to it, and that is likely due to it's collision box being too large.

If you enable debug printing (`tb.Ballistics.PrintBallisticsDebugInfo 1`), you may be able to see some more detailed information in the output log.
